components/python/python-312/patches/44-disable_epoll.patch

```diff
@@ -0,0 +1,22 @@
+
+Without this patch, python detects and uses epoll which only exists in
+OmniOS for lx zones and Linux compatibility. It is not quite the same as
+the Linux implementation and can cause socket related failures in python.
```

Comment:
This is something that needs some love:

- we are not OmniOS :-)
- we do not have lx zones
- it is strange that OmniOS claims (this patch was apparently copied from OmniOS) that since epoll is for Linux compatibility then it should not be used.  And the reason is that our implementation is not quite the same as on Linux.  The [epoll support](https://www.illumos.org/issues/5640) was added to be fully compatible with Linux and to make porting of software from Linux easier.  If there is some notable difference, then it is a bug in illumos that should be (at least) reported, and then investigated and eventually fixed.
- it should be checked/tested if this patch is still needed;  if it is still needed then it means that something in illumos epoll does not work as expected (see above), or there is some bug in Python itself.  Either way we need to report this as a bug.

Thanks.

Reply:
The epoll support was not intended to be fully compatible as some behaviours which were considered as buggy in the Linux implementation were not replicated (there was a presentation by Cantrill along these lines), therefore illumos epoll can only be used reliably in a subset of the situations encountered. I cannot remember the details but there are situations where the Linux epoll "works" which should be considered as undefined behaviour, like referencing resources that are long gone. Mimicing Linux's behaviour would rely on border effects. In such cases disabling epoll was the only reasonable way to go. The good news is that usually epoll would just hang on illumos when such edge cases are met.

Reply:
For the same reason epoll is disabled in Xorg and other graphical toolkits.

Reply:
Okay, understood.  But still we need better comment here in this patch.  Reference to more info would be great.  For exactly the reason you stated: "I cannot remember..."

Also, hang in epoll (is the hang in illumos or in the epoll consumer?) cannot be considered as good news.  It sounds like a bug.

Reply:
It looks like there are many problems related to epoll in Python itself.

Examples:
- on Linux: https://github.com/python/cpython/issues/112741
- on Windows (yes, they use epoll even on Windows!): https://github.com/python/cpython/issues/111604

But I found only one (long time ago solved) issue reported there related to illumos and epoll:
- https://github.com/python/cpython/issues/73572

So we definitely need more bug reports :-).

Reply:
It looks like the [epoll(7) man page](https://illumos.org/man/7/epoll) contains all needed information, including caveats about Linux epoll implementation.